observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java

```diff
@@ -0,0 +1,436 @@
+          .setPayloadSize(pair.size)
+          .setRpcId(rpcId);
+      if (peerAddress != null) {
+        logEntryBuilder.setPeerAddress(socketAddressToProto(peerAddress));
+      }
+      logger.log(Level.FINEST, "Writing Response Header to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+
+    /**
+     * Logs the server trailer.
+     */
+    void logTrailer(
+        long seqId,
+        String serviceName,
+        String methodName,
+        Status status,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(status, "status");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_CLIENT,
+          "peerAddress can only be specified for client");
+
+      PayloadBuilder<GrpcLogRecord.Metadata.Builder> pair = createMetadataProto(metadata);
+      GrpcLogRecord.Builder logEntryBuilder = createTimestamp()
+          .setSequenceId(seqId)
+          .setServiceName(serviceName)
+          .setMethodName(methodName)
+          .setEventType(EventType.GRPC_CALL_TRAILER)
+          .setEventLogger(eventLogger)
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setMetadata(pair.proto)
+          .setPayloadSize(pair.size)
+          .setStatusCode(status.getCode().value())
+          .setRpcId(rpcId);
+      String statusDescription = status.getDescription();
+      if (statusDescription != null) {
+        logEntryBuilder.setStatusMessage(statusDescription);
+      }
+      byte[] statusDetailBytes = metadata.get(STATUS_DETAILS_KEY);
+      if (statusDetailBytes != null) {
+        logEntryBuilder.setStatusDetails(ByteString.copyFrom(statusDetailBytes));
+      }
+      if (peerAddress != null) {
+        logEntryBuilder.setPeerAddress(socketAddressToProto(peerAddress));
+      }
+      logger.log(Level.FINEST, "Writing Server Trailer to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+
+    /**
+     * Logs the RPC message.
+     */
+    <T> void logRpcMessage(
+        long seqId,
+        String serviceName,
+        String methodName,
+        GrpcLogRecord.EventType eventType,
+        T message,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          eventType == EventType.GRPC_CALL_REQUEST_MESSAGE
+              || eventType == EventType.GRPC_CALL_RESPONSE_MESSAGE,
+          "event type must correspond to client message or server message");
+      checkNotNull(message, "message");
+
+      // TODO(dnvindhya): Convert message to bystestring
+      // byte[] messageArray = (byte[])message;
```

Comment:
You can't just typecast message to `byte[]`. But you should be able to call `.toByteString()` (or `.toByteArray()` in case you really want a byte array and not a ByteString`) on message because a message is basically a  `com.google.protobuf.Message` (i.e. an instance of a subclass of `com.google.protobuf.Message`) so you don't really need a byte-marshaller or such thing